### ISSUE TITLE
Net.scala does not seem to learn. For instance, the reported "ll" value keeps alternating between two values.

### ISSUE THREAD

**Issue by DanielTakeshi** [BIDData/BIDMach#95]
BIDMach version: 9ee9e43cfac02ee17a526fd63b8bb96790cd0b57 (i.e., as of yesterday)
BIDMat version: aec11c71bd1e785b1310886ff88477d4f26a5640 (i.e., as of yesterday)
Computer: Linux ("bitter", actually)

I'm using the Net.scala class for regression and so far I'm unable to get the network to learn anything meaningful. I am using the following four data matrices, which are in the `florian_data` directory on my bitter directory:

```
grasp_data_00.fmat.lz4
grasp_data_01.fmat.lz4
grasp_target_00.fmat.lz4  
grasp_target_01.fmat.lz4 
```

Where the two 'data' matrices have 1598 rows (i..e, features), and around 120k columns (i.e., data instances), and the two 'target' matrices have just one row. The target values are all between [0,1] so the network is performing regression.

I am using the following script to run BIDMach's Net, which is based on the `scripts/testnet.ssc` script. I am calling it `scripts/testnet_daniel.ssc`. I also added some comments for readability. (Best viewed on Github rather than email.)

```
import BIDMach.networks.Net
val dir = "florian_data/"

// Training data
val a0 = loadFMat(dir + "grasp_data_00.fmat.lz4")
val c0 = loadFMat(dir + "grasp_target_00.fmat.lz4")
val rr = rand(c0.ncols,1);
val (ss, ii) = sort2(rr);
val train_features_shuffled = a0(?,ii);
val train_targets_shuffled = c0(?,ii);

// The test data, which we predict on. Use one file number for this, I suppose.
val test_features = loadFMat(dir + "grasp_data_01.fmat.lz4")(?, 0->123000)
val test_targets = loadFMat(dir + "grasp_target_01.fmat.lz4")(?, 0->123000)

// One main confusion: why two "opts" stuff? Is one for training and the other for the architecture?
val opts = new Net.LearnOptions
opts.aopts = opts
opts.batchSize = 2000
opts.reg1weight = 0.0001
opts.npasses = 1 
opts.addConstFeat = false;
opts.links = iones(1,1); // I'm *pretty* sure this should have just one row based on other examples ...
opts.lrate = 0.01f
opts.texp = 0.4f
opts.nweight = 1e-4f
opts.what

// Last parameter is an integer from 1 to 4 and represents the non-linearity.
val net = Net.dnodes(7, 1000, 0.75f, 1, opts, 2); 
val (nn,opts0) = Net.learnerX(train_features_shuffled, train_targets_shuffled);
opts0.nodeset = net 
opts0.npasses = 100 
opts0.batchSize = 10000
opts0.lrate = 0.01f
opts0.what
val dnn = nn.model.asInstanceOf[Net]
nn.train

// Now predict. Strangely, the batch size here means we ignore the last "(# cols) mod 1000" elements.
val (mm,mopts) = Net.predictor(dnn, test_features);
val dmm = mm.model.asInstanceOf[Net]
mopts.batchSize = 1000
mm.predict

val predictions = FMat(mm.preds(0))
size(predictions)
size(test_targets)
val diff = abs(predictions - test_targets)
mean(diff)
```

I changed the training and testing data so that the training and testing were files numbered "00" and "01", respectively. Notice that the batch size is 10k, and that I'm running this for 100 iterations. Finally, I am using the `dnodes` method, with 7 layers. This means the network looks like this:

1598 as input ==> 1000 fully connected layer (w/sigmoid) ==> 1000*0.75 fully connected layer (w/sigmoid) ===> output

However, when I run this script, I get the following output for the first few passes:

```
pass= 0
16.00%, ll=-0.25474, gf=1642.817, secs=0.1, GB=0.13, MB/s=1031.61, GPUmem=0.946021
100.00%, ll=0.00000, gf=2190.271, secs=0.8, GB=0.79, MB/s=979.35, GPUmem=0.944169
pass= 1
16.00%, ll=-0.66177, gf=2159.099, secs=0.9, GB=0.92, MB/s=1005.76, GPUmem=0.944169
100.00%, ll=0.00000, gf=2241.659, secs=1.6, GB=1.58, MB/s=1002.33, GPUmem=0.944169
pass= 2
16.00%, ll=-0.25483, gf=2222.833, secs=1.7, GB=1.71, MB/s=1015.81, GPUmem=0.944169
100.00%, ll=0.00000, gf=2261.256, secs=2.3, GB=2.37, MB/s=1011.09, GPUmem=0.944169
pass= 3
16.00%, ll=-0.26091, gf=2247.492, secs=2.5, GB=2.50, MB/s=1019.97, GPUmem=0.944169
100.00%, ll=0.00000, gf=2271.183, secs=3.1, GB=3.16, MB/s=1015.53, GPUmem=0.944169
pass= 4
16.00%, ll=-0.25766, gf=2260.380, secs=3.2, GB=3.29, MB/s=1022.14, GPUmem=0.944169
100.00%, ll=0.00000, gf=2278.356, secs=3.9, GB=3.95, MB/s=1018.73, GPUmem=0.944169
pass= 5
16.00%, ll=-1.22978, gf=2269.441, secs=4.0, GB=4.08, MB/s=1023.99, GPUmem=0.944169
100.00%, ll=0.00000, gf=2283.163, secs=4.6, GB=4.74, MB/s=1020.88, GPUmem=0.944169
pass= 6
16.00%, ll=-0.26548, gf=2276.058, secs=4.7, GB=4.87, MB/s=1025.46, GPUmem=0.944169
100.00%, ll=0.00000, gf=2287.031, secs=5.4, GB=5.53, MB/s=1022.61, GPUmem=0.944169
pass= 7
16.00%, ll=-1.24147, gf=2280.426, secs=5.5, GB=5.66, MB/s=1026.34, GPUmem=0.944169
100.00%, ll=0.00000, gf=2290.312, secs=6.2, GB=6.32, MB/s=1024.08, GPUmem=0.944169
```

What's interesting is that by the time we get to five passes, the reported log likelihood keeps alternating between two values. This continues to the end:

```
pass=95
16.00%, ll=-1.36909, gf=2232.413, secs=75.3, GB=75.21, MB/s=998.68, GPUmem=0.944169
100.00%, ll=0.00000, gf=2232.428, secs=76.0, GB=75.87, MB/s=998.20, GPUmem=0.944169
pass=96
16.00%, ll=-0.25814, gf=2231.849, secs=76.1, GB=76.00, MB/s=998.42, GPUmem=0.944169
100.00%, ll=0.00000, gf=2231.869, secs=76.8, GB=76.66, MB/s=997.95, GPUmem=0.944169
pass=97
16.00%, ll=-1.36689, gf=2231.326, secs=76.9, GB=76.79, MB/s=998.18, GPUmem=0.944169
100.00%, ll=0.00000, gf=2231.466, secs=77.6, GB=77.45, MB/s=997.77, GPUmem=0.944169
pass=98
16.00%, ll=-0.25805, gf=2230.957, secs=77.7, GB=77.58, MB/s=998.02, GPUmem=0.944169
100.00%, ll=0.00000, gf=2231.184, secs=78.4, GB=78.24, MB/s=997.64, GPUmem=0.944169
pass=99
16.00%, ll=-1.36408, gf=2230.681, secs=78.5, GB=78.37, MB/s=997.89, GPUmem=0.944169
100.00%, ll=0.00000, gf=2230.964, secs=79.2, GB=79.03, MB/s=997.54, GPUmem=0.944169
Time=79.2290 secs, gflops=2230.94
```

The predictions aren't really that helpful. I've printed some of them out (same to three decimal places, I think):

```
predictions: BIDMat.FMat = 0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10323,0.10313,0.10313,0.10313,0.10313,0.10314,0.10322,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10322,0.10313,0.10313,0.10313,0.10313,0.10322,0.10321,0.10313,0.10313,0.10313,0.10323,0.10313,0.10313,0.10313,0.10322,0.10315,0.10313,0.10322,0.10321,0.10313,0.10313,0.10323,0.10313,0.10313,0.10312,0.10313,0.10313,0.10312,0.10313,0.10323,0.10313,0.10313,0.10322,0.10323,0.10313,0.10316,0.10320,0.10313,0.10319,0.10313,0.10313,0.10323,0.10319,0.10313,0.10313,0.10313,0.10323,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10322,0.10313,0.10313,0.10313,0.10318,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10313,0.10312,0.10313,0.10312,0.10313,0.10313,0.10313,...
```

Here's what I've tried:

1. Using `dnodes4` instead of `dnodes`
2. Using 3 or 5 layers instead of 7
3. Changing the batch size of opts0 (not opts). There are about 120k training instances so the batch size should probably be on the order of maybe 10k to have one significant held-out set.
4. Using the `opts0.hasBias` parameter.
5. Changing the learning rate of both opts and opts0 (there's two of them). For opts.lrate the default was 0.4 and I tried 0.01, 0.001, etc., but that seems to change the magnitude of predictions, not the accuracy. For the opts0.lrate, the default is one and I tried 0.01f.

I suspect this is an issue with the learning rate, given that Python's theano library can learn with this data. Do you have any suggestions on how to diagnose this?

Thanks,
Daniel

**Comment by jcanny** [comment]
Hi Daniel,
Please mail me first if you have questions about the Learner. It doesnt 
make sense to post issues until you're sure that they're issues with the 
toolkit vs learning how to use it.

-John

**Comment by jcanny** [comment]
Your script is setting the minibatch size to 100000 (almost the whole dataset), and its creating the net before the options are set. I put a cleaned up sample script in testnet.ssc. It works fine on that dataset and is quite insensitive to learning rate. We havent released the Net code yet and its not fully documented but it doesnt make sense to post issues related to its use.

See also testgrasp.ssc which runs GLM, Net and Random Forests on the grasp data.

**Comment by DanielTakeshi** [comment]
Thanks John, it works now.

By the way, there might be a slight confusion, given that in this previous comment (https://github.com/BIDData/BIDMach/issues/50) you mention that DNN is being replaced by the Net class. I assume that neither were completed but that the Net is the most up to date, so if there are any future issues I will report for Net.scala.

Thanks again.

**Comment by DanielTakeshi** [comment]
PS: Would it be possible if you could add testgrasp.ssc to the scripts directory? It does not seem to be there.
